### ISSUE TITLE
Specifiying a non-Vim Python "distribution" using absolute path, for Windows OS?

### ISSUE THREAD

**Issue by llinfeng** [python-mode/python-mode#884]
# Question
Can I specify the absolute path for Python3? 

I learned from the documentation that I may change the interpreter to `Python3` by setting `let g:pymode_python = 'python3'`. Yet, I don't find a way to point Vim to the correct interpreter on my Windows machine. The workhorse Python interpreter is found as `c:\ProgramData\Anaconda3\python.exe` on my system. Please advise how to specify this "python version" to `python-mode`. 

# Attempts and clarification
First, I **need** to fix the "Vim-Python" version, as `let $PYTHONHOME = 'c:\Python27\'`. While this is a Python distribution that I never actively use, a good number of my Vim plugins are running on this distribution (Python 2.14, to be precise). These plugin include: `vim-tex`, `VOom` and so on.

I have also tried to create a `python3` shortcut in my `PATH`, which is not yet available to my command-prompt (nor Powershell), but should be callable through the Windows Run window. This does not point `python-mode` to look for my active installation as found as: `c:\ProgramData\Anaconda3\python.exe`. 

# Confusion
What is `python-mode` looking for when I set `let g:pymode_python = 'python3'`? Is it some command-line executable named as `python3`?

## Python 3.5, or my system "`Python3`"?
It confused me when seeing the following error message after attempting to use Pyhton3. `python-mode` seems to be looking for `python 3.5` by *default*? My workhorse Python distribution is from Anaconda, which I have made available to my command-line (Windows Command Prompt) as simply `python`.
![pymode_python_version](https://user-images.githubusercontent.com/5890407/39363032-c675b3b8-49f6-11e8-8ecc-773bae74d35e.jpg)

-Thank you in advance, Linfeng.

**Comment by diraol** [comment]
@llinfeng _as far as I understand_, yes, by using `let g:pymode_python = 'python3'` you are saying to vim to look for a `python3` executable.

If you have made your Anaconda Python3 available as `python`, then you may just use `let g:pymode_python = 'python'`.

_I am not sure if it will work_, but you can also try to do `let g:pymode_python = 'c:\ProgramData\Anaconda3\python.exe'`.

Using `let g:pymode_python = 'python3'` usually works on systems that have both python2 and python3, but defaults `python` to the `python2` version and `python3` to `python3`.

**Comment by llinfeng** [comment]
@diraol So far, I have tried the following three ways of pointing Pymode to my *Windows* Python:
```
let g:pymode_python = 'C:\[absolute-path]\Anaconda3\python.exe'
let g:pymode_python = 'python3'
let g:pymode_python = 'python'
```
Among these settings, `python` pointed me to the _'Vim-default" Python 2.7.14_, and is the only option that have had `python-mode` to load; `python3` and the absolute path both gave me the exact same error messages from my original post. 

## "Vim-specific" *`python`*?
I am not sure how it happened, but firing `:!python` through command-line of GVim does point me to my Python 2.7 distribution. This explains why specifying `let g:pymode_python = 'python'` will correctly point `python-mode` to Python 2.7, which worked.

However, the following screenshot is _confusing_: w.r.t. my command prompt, both `python` and `python3` are pointing to the Anaconda-managed Python 3.6
![system_python_through_command promp](https://user-images.githubusercontent.com/5890407/46215337-1fc71f00-c30b-11e8-9148-3a68c99502b5.png)
I think this is achieved by me adding the absolute-path-folder of Anaconda-Python3 to **the top of** my system PATH.

## Is my Vim-on-Windows distribution bad? a.k.a. How `python35.dll` was attempted for loading?
Up to this point, I doubt if it is a build problem: through `:version`, I do find both `+python/dyn` and `+python3/dyn`; but I was only able to specify the following setting through my `.vimrc`:
```
let $PYTHONHOME = 'c:\Python27\'
```
This appears to be a justification for why `:!python` tries to get me Python 2.7, and that `python-mode` had trouble loading `python35.dll`?? (Well, I am not quite sure where this `python35` come from.)
 
**Vim-info**: well, Gvim from https://www.vim.org/download.php#pc
![vim_version](https://user-images.githubusercontent.com/5890407/46215792-42a60300-c30c-11e8-9586-c98a90ed716f.png)


## Will specifying `let g:pymode_virtualenv_path="something"` help?
I haven't got virtual environemnts to work well, but I can give it a try if this should be an opportunity to solve this long-standing problem.

**Comment by llinfeng** [comment]
## A new error message using gvim81.exe

![new_python_error](https://user-images.githubusercontent.com/5890407/46218355-55233b00-c312-11e8-90fc-dc218b470942.png)

This error is occurring in the following distributions of gvim81:
* Directly downloaded from <vim.org>: https://www.vim.org/download.php#pc
* Downloaded through Chocolatey: https://chocolatey.org/packages/vim
    * Well, this particular distribution is even worse: leaving the setting to `python` won't help much.

I am reverting back to my previous vim80 on Windows, and will try to get things running through the Windows Subsystem for Linux.

**Comment by jueqingsizhe66** [comment]
> ## A new error message using gvim81.exe
> ![new_python_error](https://user-images.githubusercontent.com/5890407/46218355-55233b00-c312-11e8-90fc-dc218b470942.png)
> 
> This error is occurring in the following distributions of gvim81:
> 
> * Directly downloaded from <vim.org>: https://www.vim.org/download.php#pc
> * Downloaded through Chocolatey: https://chocolatey.org/packages/vim
>   
>   * Well, this particular distribution is even worse: leaving the setting to `python` won't help much.
> 
> I am reverting back to my previous vim80 on Windows, and will try to get things running through the Windows Subsystem for Linux.

I have the same error message

**Comment by llinfeng** [comment]
...
> I have the same error message

@jueqingsizhe66 I am inclined to give up writing Python scripts using the "Windows-native" Vim ==> up to this time, I can hardly use Gvim on Windows to get `python-mode` to work properly. 

**Solution**: the WSL is good enough and is easy to set up. I will try to see if I can get everything to work there. (WSL stands for Windows subsystem for Linux, and one can run Linux kernels from the Windows machine. I am using ConEmu to access it, and things are so far so good except for temporary display issues. This is overcome by using `tmux` at all time: when there is a display error, detaching the session and attaching again will help solve the display issue.)

@diraol Is there a quick setup instruction that deals with the following installation steps:
* Get a fully Python-friendly Vim for Unix (please identify a good enough "build source"?) Namely, with `+python` and `+python3` enabled.
* Get a `python-mode` friendly Python(3) "Kernel"? Or, any base version of Python3 shall work. (I should be using Python3 exclusively, but some of my Vim plugins do need to access Python 2 through `python`.)

**Comment by diraol** [comment]
@llinfeng As far as I remember Vim can't handle both python2 and python3 at the same time. I don't remember why, but you need to choose one. If you need both, you should build and install one version for each (and have something like vim2 and vim3).
Regarding `python-mode`, despite some bugs, it should work fine with both python2 and 3 (even 3.7).

About "Python-friendly Vim for Unix", you can try using "vim-nox" or build your own from the vim source code (their instructions are quite nice, but you can look for "building vim for debian" or something like that).

**Comment by llinfeng** [comment]
@diraol Thank you for clarifying the intended usage. Switching to `vim-nox` does help. I no longer get any buggy things **after switching to WSL** on my Windows 10 machine.

I found the following change-log [here](https://raw.githubusercontent.com/vim/vim/master/runtime/doc/version8.txt). Does the option `python` or `python3` for variable `g:pymode_python` have anything to do with the `$PYTHONHOME` environmental variable? Or, it talks to the terminal and tries to evaluate `python3`, for example, as a standalone command? (Well, the true question ==> Implicitly, changing the alias in `~/.bashrc` will empower people to point the `pymode_python` to other python distribution that has been installed?)

> Patch 8.0.1451
> Problem:    It is difficult to set the python home directory properly for
>             Python 2.7 and 3.5 since both use $PYTHONHOME.
> Solution:   Add the 'pythonhome' and 'pythonthreehome' options. (Kazuki
>             Sakamoto, closes #1266)
> Files:      runtime/doc/options.txt, runtime/doc/quickref.txt,
>             runtime/optwin.vim, src/if_python.c, src/if_python3.c,
>             src/option.c, src/option.h

Following such pair of new environmental variables, it may improve the situation that "Vim can't handle both python2 and python3"?
